Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift

```diff
@@ -184,3 +187,109 @@ public final class AesKw: NSObject {
+            out: &d,
+            Insecure.SHA1.self
+        )
+    }
+    @objc public static func sha256(
+        _ data: UnsafePointer<UInt8>, dataSize: UInt, digest: UnsafeMutablePointer<UInt8>,
+        digestSize: UInt
+    ) -> ErrorCodes {
+        guard
+            checkInputs(dataSize: dataSize, digestSize: digestSize, SHA256.self) == .success
+                && digestSize == SHA256.byteCount
+        else {
+            return .invalidArgument
+        }
+        var d = UnsafeMutableRawBufferPointer(start: digest, count: Int(digestSize))
+        return Self.digest(
+            data: UnsafeBufferPointer(start: data, count: Int(dataSize)), out: &d, SHA256.self)
+    }
+    @objc public static func sha384(
+        _ data: UnsafePointer<UInt8>, dataSize: UInt, digest: UnsafeMutablePointer<UInt8>,
+        digestSize: UInt
+    ) -> ErrorCodes {
+        guard
+            checkInputs(dataSize: dataSize, digestSize: digestSize, SHA384.self) == .success
+                && digestSize == SHA384.byteCount
+        else {
+            return .invalidArgument
+        }
+
+        var d = UnsafeMutableRawBufferPointer(start: digest, count: Int(digestSize))
+        return Self.digest(
+            data: UnsafeBufferPointer(start: data, count: Int(dataSize)), out: &d, SHA384.self)
+    }
+    @objc public static func sha512(
+        _ data: UnsafePointer<UInt8>, dataSize: UInt, digest: UnsafeMutablePointer<UInt8>,
+        digestSize: UInt
+    ) -> ErrorCodes {
+        guard
+            checkInputs(dataSize: dataSize, digestSize: digestSize, SHA512.self) == .success
+                && digestSize == SHA512.byteCount
+        else {
+            return .invalidArgument
+        }
+
+        var d = UnsafeMutableRawBufferPointer(start: digest, count: Int(digestSize))
+        return Self.digest(
+            data: UnsafeBufferPointer(start: data, count: Int(dataSize)), out: &d, SHA512.self)
+    }
+    private static func digest<T: HashFunction>(
+        data: UnsafeBufferPointer<UInt8>, out: inout UnsafeMutableRawBufferPointer, _: T.Type
+    ) -> ErrorCodes {
+        var hasher = T()
+        var inputData = data
+        if data.count == 0 {
+            Utils.zeroArray.withUnsafeBufferPointer { buf in
+                inputData = buf
+            }
+        }
+        hasher.update(
+            bufferPointer: UnsafeRawBufferPointer(
+                start: inputData.baseAddress, count: inputData.count))
+        let result = hasher.finalize()
+        return result.withUnsafeBytes { buf in
+            if buf.count != out.count {
+                return ErrorCodes.hashingFailed
+            }
+            out.copyBytes(from: buf)
+            return .success
+        }
+    }
+    private static func checkInputs<T: HashFunction>(dataSize: UInt, digestSize: UInt, _: T.Type)
+        -> ErrorCodes
+    {
+        if dataSize > Int.max
+            || digestSize == 0
+            || digestSize > Int.max
```

Comment:
These seem safely impossible.  We don't control dataSize, but we do control these.

Reply:
they do but then across the boundary all bets are off and I think the performance overhead is probably worth it for future.